docs/design.md

```diff
@@ -0,0 +1,382 @@
+
+**Example**
+
+    {
+      "messageType": "hello",
+      "uaid": "fd52438f-1c49-41e0-a2e4-98e49833cc9c"
+    }
+
+### Register
+
+The Register message is used by the UserAgent to request that the PushServer
+notify it when a channel changes. Since channelIDs are associated with only one
+UAID, this effectively creates the channel, while unregister destroys the
+channel.
+
+The channelID is chosen by the UserAgent because it also acts like a nonce for
+the Register message itself. Because of this PushServers MAY respond out of
+order to multiple register messages or messages may be lost without compromising
+correctness of the protocol.
+
+The request is considered successful only after a response is received with a
+status code of 200. On success the UserAgent MUST:
+
+* Update its persistent storage based on the response
+* Notify the application of a successful registration.
+* On error, the UserAgent MUST notify the application as soon as possible.
+
+**NOTE**: The register call is made by the UserAgent on behalf of an
+application. The UserAgent SHOULD have reasonable timeouts in place so that
+the application is not kept waiting for too long if the server does not
+respond or the UserAgent has to retry the connection.
+
+#### UserAgent -> PushServer
+
+messageType = "register"
+
+channelID string REQUIRED
+: A unique identifier generated by the UserAgent, distinct from any existing
+  channelIDs it has registered. It is RECOMMENDED that this is a UUIDv4 token.
+
+**Example**
+
+    {
+        "messageType": "register",
+        "channelID": "d9b74644-4f97-46aa-b8fa-9393985cd6cd"
+    }
+
+#### PushServer -> UserAgent
+
+messageType = "register"
+
+channelID string REQUIRED
+: This MUST be the same as the channelID sent by the UserAgent in the register
+  request that this message is a response to.
+
+status number REQUIRED
+: Used to indicate success/failure. MUST be one of:
+
+  * 200 - OK. Success. Idempotent: If the PushServer receives a register for the
+    same channelID from a UserAgent which already has a registration for the
+    channelID, it should still respond with success.
+
+  * 409 - Conflict. The chosen ChannelID is already in use and NOT associated
+    with this UserAgent. UserAgent SHOULD retry with a new ChannelID as soon as
+    possible.
+
+  * 500 - Internal server error. Database out of space or offline. Disk space
+    full or whatever other reason due to which the PushServer could not grant
+    this registration. UserAgent SHOULD avoid retrying immediately.
+
+pushEndpoint string REQUIRED
+: Should be the URL sent to the application by the UserAgent. AppServers will
+  contact the PushServer at this URL to update the version of the channel
+  identified by channelID.
+
+**Example**
+
+     {
+       "messageType": "register",
+       "channelID": "d9b74644-4f97-46aa-b8fa-9393985cd6cd",
+       "status": 200,
+       "pushEndpoint": "http://pushserver.example.org/d9b74644"
+     }
+
```

Comment:
Bridge functions like GCM may include additional bits of data, thus probably not a good idea to create a strict YAML for this. If that's a problem we could put the data in an "extra" block that can be ignored by strict interpreters (e.g. Go, Java, etc). Thoughts?